manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,451 @@
+	// Commit will always succeed and does not need to return an error.
+	Commit()
+
+	// IsNoop returns true if the Proposal doesn't actually modify the net
+	// allocator state. IsNoop does not mean the ports for an endpoint haven't
+	// changed. It only means that the publish ports marked in use by the
+	// allocator haven't changed.
+	//
+	// Used mostly for testing, and should probably not be relied on outside of
+	// this package.
+	IsNoop() bool
+}
+
+type proposal struct {
+	pa         *Allocator
+	ports      []*api.PortConfig
+	allocate   map[port]struct{}
+	deallocate map[port]struct{}
+}
+
+func (prop *proposal) Ports() []*api.PortConfig {
+	if prop.ports == nil {
+		return []*api.PortConfig{}
+	}
+	return prop.ports
+}
+
+// Commit commits the proposal to the port allocator.
+func (prop *proposal) Commit() {
+	if prop.IsNoop() {
+		// nothing to do if proposal is noop, short circuit a bit
+		return
+	}
+	// The pattern here is we're going to free every port in p.deallocate and
+	// then allocate every port in allocate. any over lap results in no net
+	// change
+	for p := range prop.deallocate {
+		delete(prop.pa.ports, p)
+	}
+	for p := range prop.allocate {
+		prop.pa.ports[p] = struct{}{}
+	}
+}
+
+// IsNoop returns true if the ports in use before this proposal are the same as
+// the ports in use after.
+func (prop *proposal) IsNoop() bool {
+	// if the allocate and deallocate sets are the same length, and every port
+	// in allocate is also found in deallocate, then this proposal is a noop
+	if len(prop.allocate) != len(prop.deallocate) {
+		return false
+	}
+	for p := range prop.allocate {
+		if _, ok := prop.deallocate[p]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// NewAllocator returns a new instance of the Allocator object
+func NewAllocator() *Allocator {
+	return &Allocator{
+		ports: make(map[port]struct{}),
+	}
+}
+
+// Restore adds the current endpoints to the local state of the port allocator
+// but does not perform any new allocation.
+func (pa *Allocator) Restore(endpoints []*api.Endpoint) {
+	// NOTE(dperny) we can be sure that we're not allocating new or conflicting
+	// state because if an endpoint is unallocated, it will not have any ports.
+	// we can't look at the Spec in this method, because the spec isn't real
+	// state, it's just what the user wants. no matter what changes, that
+	// invariant needs to hold.
+
+	// create and commit a proposal for port allocation
+	prop := &proposal{
+		pa: pa,
+		// we only allocate
+		allocate: make(map[port]struct{}),
+	}
+	for _, endpoint := range endpoints {
+		for _, p := range endpoint.Ports {
+			// ignore host-mode ports
+			if p.PublishMode != api.PublishModeHost {
+				prop.allocate[port{p.Protocol, p.PublishedPort}] = struct{}{}
+			}
+		}
+	}
+	prop.Commit()
+}
+
+// Deallocate takes an endpoint and provides a Proposal that will deallocate
+// all of its ports.
+func (pa *Allocator) Deallocate(endpoint *api.Endpoint) Proposal {
```

Comment:
Should this function admit variadic arguments? `func (pa *Allocator) Deallocate(endpoints ...*api.Endpoint) Proposal {` Would only need using a `for` loop wrapping the current one and will allow current behaviour but would optimize deallocating multiple endpoints by grouping them in a single `Proposal`. May you ever need multiple endpoints?

Reply:
Nah, because you very rarely are deallocating many endpoints at once. You're usually deallocating them one-by-one. To make use of a variadic, you'd have to write the caller to do batching. It's not worthwhile.

Reply:
Understood